File: Algorithms/238. Product of Array Except Self/Solution.py
#coding=utf-8
__author__ = 'xuxuan'
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        product=1
        zeroNum=0;
        zeroIndex=-1;
        for i in range(len(nums)):
            if nums[i]==0:
                zeroNum+=1
                zeroIndex=i
                if zeroNum>1:
                    break
            else :
                product*=nums[i]
        ans=[0]*len(nums)
        if zeroNum==0:
            for i in range(len(nums)):
                ans[i]=product/nums[i]
        elif zeroNum==1:
            ans[zeroIndex]=product
        return ans

File: Algorithms/238. Product of Array Except Self/Solution.java
public class Solution {
    public int[] productExceptSelf(int[] nums) {
        int length=nums.length;
        int k=0;
        boolean flag=true;
        int product=1;
        for (int i = 0; i < length; i++) {
            if (nums[i]==0) {
                if (flag) {
                    k=i;
                    flag=false;
                }else {
                    int[] answer=new int[length];
                    for (int j = 0; j <length; j++) {
                        answer[j]=0;
                    }
                    return answer;
                }
            }else {
                product*=nums[i];
            }
        }
        int[] answer=new int[length];
        if (flag) {
            for (int i = 0; i < length; i++) {
                answer[i]=product/nums[i];
            }
        }else {
            answer[k]=product;
        }
        return answer;
    }
}